### PR TITLE
[feature-vpt-integration-1674] Add whitelist functionality

### PR DESCRIPTION
**Pull request by markeissler**
This PR is part of the following series of sequentially dependent (stacked) branches. Please review and comment on changes in each PR individually.

**NOTE 1: BRANCHES MUST be merged in reverse order (1677-->1521) to avoid conflicts.**

**NOTE 2: PR number order and merge order are not necessarily the same. Verify the name of the PR before merging into its parent branch.**

The stack includes (sequentially, in reverse):

- feature-vpt-integration-1677
- feature-vpt-integration-1675
- feature-vpt-integration-1674
- feature-vpt-integration-1521-1
- feature-vme-integration-1521

This specific PR adds support for whitelist functionality:

- Add DomainWhitelist and DomainWhitelistFile config options
- Loads whitelist file contents using existing parsing routines
- Copies whitelist file contents to a hash for efficiency
- Updates man page for DomainWhitelist and DomainWhitelistFile configs

### CODE REVIEW

File: opendmarc/opendmarc.c [hunk 1]
```diff
@@ -3017,30 +3066,32 @@ mlfi_eom(SMFICTX *ctx)
...
+	ret = SMFIS_ACCEPT;
 
 	switch (policy)
 	{
 	  case DMARC_POLICY_ABSENT:		/* No DMARC record found */
 	  case DMARC_FROM_DOMAIN_ABSENT:	/* No From: domain */
 		aresult = "none";
-		ret = SMFIS_ACCEPT;
-		result = DMARC_RESULT_ACCEPT;
+		// ret = SMFIS_ACCEPT;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Just remove the lines if we don't want them anymore.

File: opendmarc/opendmarc.c [hunk 2]
```diff
@@ -3017,30 +3066,32 @@ mlfi_eom(SMFICTX *ctx)
...
+		// result = DMARC_RESULT_ACCEPT;
 		break;
 
 	  case DMARC_POLICY_NONE:		/* Alignment failed, but policy is none: */
 		aresult = "fail";		/* Accept and report */
-		ret = SMFIS_ACCEPT;
-		result = DMARC_RESULT_ACCEPT;
+		// ret = SMFIS_ACCEPT;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Remove.

File: opendmarc/opendmarc.c [hunk 3]
```diff
@@ -3017,30 +3066,32 @@ mlfi_eom(SMFICTX *ctx)
...
+		// result = DMARC_RESULT_ACCEPT;
 		break;
 
 	  case DMARC_POLICY_PASS:		/* Explicit accept */
 		aresult = "pass";
-		ret = SMFIS_ACCEPT;
-		result = DMARC_RESULT_ACCEPT;
+		// ret = SMFIS_ACCEPT;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Remove.

File: opendmarc/opendmarc.c [hunk 4]
```diff
@@ -1248,6 +1269,34 @@ dmarcf_config_load(struct config *data, struct dmarcf_config *conf,
 			                  sizeof conf->conf_dnstimeout);
 		}
 
+		str = NULL;
+		(void) config_get(data, "DomainWhitelist", &str, sizeof str);
+		if (str != NULL)
+		{
+			ENTRY entry, *entryptr;
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Single lines, please.

File: opendmarc/opendmarc.c [hunk 5]
```diff
@@ -1248,6 +1269,34 @@ dmarcf_config_load(struct config *data, struct dmarcf_config *conf,
...
+			int result = 0;
+
+			dmarcf_mkarray(str, &conf->conf_domainwhitelist);
+			/* add domains to hash */
+			for(int i = 0; conf->conf_domainwhitelist[i] != NULL; i++)
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Space after "for".

File: opendmarc/opendmarc.c [hunk 6]
```diff
@@ -3670,6 +3720,16 @@ main(int argc, char **argv)
 	conffile = NULL;
 	ignore = NULL;
 
+	/* init domain_whitelist_hash table */
+	domain_whitelist_hash = calloc(1, sizeof(struct hsearch_data));
+	if (hcreate_r(MAXWHITELISTSIZE, domain_whitelist_hash) == 0)
+	{
+		fprintf(stderr, "%s: malloc(): %s\n", progname,
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Error message should reflect what call failed.  It wasn't malloc() here.

File: opendmarc/opendmarc.c [hunk 7]
```diff
@@ -3972,10 +4032,66 @@ main(int argc, char **argv)
...
+
+#if defined(__linux__) && defined(DEBUG_WHITELIST)
+	/* walk through the hash and print keys and values */
+	struct hsearch_data *hdp = domain_whitelist_hash;
+
+	fprintf(stderr, "domain_whitelist_hash contents...\n");
+	for (int i = 0; i < hdp->size; i++)
+	{
+		if (hdp->table[i].used)
+		{
```
Comment (mskucherawy) [inline opendmarc/opendmarc.c]: Braces not needed for a one-line block.

**Comment by markeissler** [comment]
Closing for collapsed stack:

* [PR#15 Add support for selectors, domain whitelist, arc.chain #15](https://github.com/trusteddomainproject/OpenDMARC/pull/15)

**Pull request closed without merging**